File: apps/mobile/modules/safely-push-content/ios/PushContentCore/PushContentCore.swift
import Foundation

public struct RewrittenPushContent: Equatable {
    public let title: String
    public let body: String
}

public enum PushContentCore {
    public static let appGroupId = "group.com.safely.wallet"
    public static let walletNamePlaceholder = "{{wallet_name}}"

    static let walletNamesFileName = "push-wallet-names.json"
    static let targetRefKey = "target_ref"
    static let titleTemplateKey = "title_template"
    static let bodyTemplateKey = "body_template"

    public static func walletNamesURL(in containerURL: URL) -> URL {
        containerURL.appendingPathComponent(walletNamesFileName, isDirectory: false)
    }

    public static func save(walletNames: [String: String], to url: URL) throws {
        let data = try JSONSerialization.data(withJSONObject: walletNames, options: [.sortedKeys])
        var options: Data.WritingOptions = [.atomic]
        #if os(iOS)
        options.insert(.completeFileProtectionUntilFirstUserAuthentication)
        #endif
        try data.write(to: url, options: options)

        var resourceURL = url
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try? resourceURL.setResourceValues(values)
    }

    public static func loadWalletNames(from url: URL) -> [String: String] {
        guard let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data),
              let names = json as? [String: String] else {
            return [:]
        }
        return names
    }

    public static func render(template: String, walletName: String) -> String? {
        let rendered = template.replacingOccurrences(of: walletNamePlaceholder, with: walletName)
        guard !containsPlaceholder(rendered) else { return nil }
        return rendered
    }

    public static func rewrite(
        title: String,
        body: String,
        data: [String: Any],
        walletNames: [String: String]
    ) -> RewrittenPushContent? {
        guard let targetRef = data[targetRefKey] as? String,
              let walletName = walletNames[targetRef] else {
            return nil
        }
        let titleTemplate = data[titleTemplateKey] as? String
        let bodyTemplate = data[bodyTemplateKey] as? String
        guard titleTemplate != nil || bodyTemplate != nil else { return nil }

        guard let newTitle = titleTemplate.map({ render(template: $0, walletName: walletName) }) ?? title,
              let newBody = bodyTemplate.map({ render(template: $0, walletName: walletName) }) ?? body else {
            return nil
        }
        return RewrittenPushContent(title: newTitle, body: newBody)
    }

    private static func containsPlaceholder(_ text: String) -> Bool {
        guard let open = text.range(of: "{{") else { return false }
        return text[open.upperBound...].contains("}}")
    }
}

File: apps/mobile/modules/safely-push-content/android/src/main/java/expo/modules/safelypushcontent/PushContentCore.kt
package expo.modules.safelypushcontent

import java.io.File
import java.util.Properties

internal data class RewrittenPushContent(val title: String, val body: String)

internal const val WALLET_NAME_PLACEHOLDER = "{{wallet_name}}"
internal const val TARGET_REF_KEY = "target_ref"
internal const val TITLE_TEMPLATE_KEY = "title_template"
internal const val BODY_TEMPLATE_KEY = "body_template"
internal const val WALLET_NAMES_FILE_NAME = "push-wallet-names.properties"

internal class WalletNameStore(private val file: File) {
    fun save(names: Map<String, String>) {
        val properties = Properties()
        names.forEach { (ref, name) -> properties.setProperty(ref, name) }
        val tmp = File(file.parentFile, file.name + ".tmp")
        tmp.parentFile?.mkdirs()
        tmp.writer(Charsets.UTF_8).use { properties.store(it, null) }
        if (!tmp.renameTo(file)) {
            tmp.copyTo(file, overwrite = true)
            tmp.delete()
        }
    }

    fun load(): Map<String, String> {
        if (!file.exists()) return emptyMap()
        return try {
            val properties = Properties()
            file.reader(Charsets.UTF_8).use { properties.load(it) }
            properties.stringPropertyNames().associateWith { properties.getProperty(it) }
        } catch (e: Exception) {
            emptyMap()
        }
    }
}

internal fun renderPushTemplate(template: String, walletName: String): String? {
    val rendered = template.replace(WALLET_NAME_PLACEHOLDER, walletName)
    return if (containsPlaceholder(rendered)) null else rendered
}

internal fun rewritePushContent(
    title: String?,
    body: String?,
    data: Map<String, String>,
    walletNames: Map<String, String>
): RewrittenPushContent? {
    val walletName = data[TARGET_REF_KEY]?.let { walletNames[it] } ?: return null
    val titleTemplate = data[TITLE_TEMPLATE_KEY]
    val bodyTemplate = data[BODY_TEMPLATE_KEY]
    if (titleTemplate == null && bodyTemplate == null) return null

    val newTitle = titleTemplate?.let { renderPushTemplate(it, walletName) ?: return null } ?: title ?: ""
    val newBody = bodyTemplate?.let { renderPushTemplate(it, walletName) ?: return null } ?: body ?: ""
    return RewrittenPushContent(newTitle, newBody)
}

private fun containsPlaceholder(text: String): Boolean {
    val open = text.indexOf("{{")
    return open >= 0 && text.indexOf("}}", open + 2) >= 0
}
